### ISSUE TITLE
The job running on agent Azure Pipelines 2 ran longer than the maximum time of 60 minutes. 

### ISSUE THREAD

**Issue by monajalal** [MicrosoftDocs/azure-devops-docs#13305]
I am using Azure Pipeline Templates and in the `pipeline.yaml` file, I have the following and I am not sure where I can insert either of the following variables?

    `cancelTimeoutInMinutes: 180`
    `timeoutInMinutes: 180`

From the documentation in https://learn.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#timeouts it's not very clear where and how I should change the timeout in `pipeline.yaml` file?

```
$schema: https://azuremlschemas.azureedge.net/latest/pipelineJob.schema.json
type: pipeline

# <inputs_and_outputs>
inputs:
  # training_images:
  training_data:
    type: uri_folder
    mode: download # pick ro_mount, rw_mount or download
    path: azureml:FAT:1
# </inputs_and_outputs>

    

# <jobs>
settings:
  default_datastore: azureml:MYDATASTORE
  continue_on_step_failure: true

jobs:
  train:
    # cancelTimeoutInMinutes: 180
    # timeoutInMinutes: 180
    type: command
    component: file:train.yaml
    compute: azureml:pose-4node-NC24r #16GPUs w r
    resources:
      instance_count: 4    # number of nodes
    distribution:
      type: pytorch
      process_count_per_instance: 4 # number of gpus

    # NOTE: set env var if needed
    environment_variables:
      NCCL_DEBUG: "INFO" # adjusts the level of info from NCCL tests
      # NCCL_TOPO_FILE: "/opt/microsoft/ndv4-topo.xml" # Use specific topology file for A100

      # NCCL_IB_PCI_RELAXED_ORDERING: "1" # Relaxed Ordering can greatly help the performance of Infiniband networks in virtualized environments.
      # NCCL_IB_DISABLE: "1" # force disable infiniband (if set to "1")
      # NCCL_NET_PLUGIN: "none" # to force NET/Plugin off (no rdma/sharp plugin at all)
      # NCCL_NET: "Socket" # to force node-to-node comm to use Socket (slow)
      NCCL_SOCKET_IFNAME: "lo" # to force Socket comm to use eth0 (use NCCL_NET=Socket)

      # UCX_IB_PCI_RELAXED_ORDERING: "on"
      # UCX_TLS: "tcp"
      # UCX_NET_DEVICES: "eth0" # if you have Error: Failed to resolve UCX endpoint...

      CUDA_DEVICE_ORDER: "PCI_BUS_ID" # ordering of gpus

      TORCH_DISTRIBUTED_DEBUG: "DETAIL"

    inputs:
      # data inputs
      train_data: ${{parent.inputs.training_data}}

      # data loading
      batch_size: 32
      manualseed: 42
      num_workers: 5
      prefetch_factor: 4
      persistent_workers: true
      pin_memory: true
      non_blocking: false

      # model
      model_arch: "vgg16"
      model_arch_pretrained: true

      # training
      num_epochs: 1 # for testing purposes
      learning_rate: 0.0001

      # profiling
      enable_profiling: false
      # multiprocessing_sharing_strategy: "file_system" # WARNING: this can cause hang at job completion

      # Model Registrataion
      register_model_as: "MY_MODEL_NAME"


# </jobs>

```

This is due to an error I keep getting as BUILD FAILED via email:

![Screenshot from 2023-05-09 07-49-16](https://github.com/MicrosoftDocs/azure-devops-docs/assets/1892917/0123b048-5346-46da-989a-e48512766714)

Here's also how my project structure look like using MLOps Pipeline Template.

![Screenshot from 2023-05-09 07-50-34](https://github.com/MicrosoftDocs/azure-devops-docs/assets/1892917/f4d0e819-1192-4955-8cdc-6a0a466764db)

Thank you for your feedback!

The issues in this repo are for documentation feedback.

If you are submitting your documentation feedback issue directly through GitHub, please provide the URL of the article so we can route the feedback to the correct author.

Please report Azure DevOps or TFS software bugs, feedback, or questions in the [Azure DevOps Developer Community](https://developercommunity.visualstudio.com/spaces/21/index.html).

**Comment by WilliamAntonRohm** [comment]
@monajalal -- Mona, thank you for your detailed feedback.  Please consider these resources:
- [Azure DevOps Services](https://azure.microsoft.com/support/devops/) 
- [Azure DevOps on Stack Overflow](https://stackoverflow.com/questions/tagged/azure-devops)

@juliakm -- Julia,, please look into this issue.

**Comment by juliakm** [comment]
@monajalal Thanks for the detailed feedback! I think this is an issue with the underlying MLOps template and not your Azure DevOps pipeline YAML. It's confusing but the code your showing isn't an Azure DevOps pipeline, it's the YAML for MLOps. 

@WilliamAntonRohm I think that @AbeOmor would be the best person to investigate this one.

**Comment by AbeOmor** [comment]
Adding @setuc and @sdonohoo to help investigate.

**Comment by sdonohoo** [comment]
The yaml example and the file pipeline.yaml file highlighted in the project structure are both definitions for the Azure Machine Learning training pipeline, not the Azure DevOps pipeline. The AzureML pipeline should run to completion even beyond 60 minutes. This appears to be a modification of the computer vision MLOps example. That training job can take over an hour and the Azure DevOps pipeline will wait for it to complete so I think this is where the problem is. The error message supports that.

The Azure DevOps pipeline to look at is mlops/devops-pipelines/deploy-model-training-pipeline.yml. It is this Azure DevOps pipeline that creates and runs the Azure Machine Learning pipeline.

Try the following change in that pipeline file to see if it helps. Around line 31:

```
    jobs:
      - job: DeployTrainingPipeline
        timeoutInMinutes: 180 # how long to run the job before automatically cancelling
        steps:
          - checkout: self
```

**Comment by setuc** [comment]
I concur with what has been shared above. In addition, if you are using Free hosted agents have a max timeout of 60 minutes per job for a private repository and 360 minutes for a public repository. Alternative ways are to get a self-hosted agent, any Microsoft hosted agent will be limited to 360 minutes. Reference:  [Job Timeout - Azure Pipelines | Microsoft Learn](https://learn.microsoft.com/en-us/azure/devops/pipelines/troubleshooting/troubleshooting?view=azure-devops#job-time-out)

We are working on a solution to address the long-running jobs as part Azure Machine Learning (AML) Pipeline. This solution can be expected to be released in the next 3 to 4 weeks. We can share the solution here once it is ready.

**Comment by monajalal** [comment]
We created our own agent pool and then set 
```
  jobs:
    - job: DeployTrainingPipeline
      cancelTimeoutInMinutes: 0
      timeoutInMinutes: 0
```

**Comment by monajalal** [comment]
@juliakm you are right. I made a mistake trying to set these values in pipeline.yaml. I should have set them in mlops --> azureml --> devoos-pipelines --> deploy-model-training-pipeline.yml

I think the document didn't make it clear to me initially if those changes need to be done in Devops ymls or Mlops ymls. 

Also, further, as @setuc mentioned, we ended up created our own docker and pushing it to acr and use it in pool: OurPoolAgents name and it was required to have Python.
